File: examples/go/apps/minipi/approval_gate.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// ApprovalGateActor — human-in-the-loop state machine.
//
// Demonstrates: GenFSM for human-in-the-loop, durable wait (agent can be
// suspended for days), and PlexSpaces's unique capability to resume exactly
// where it left off without losing state.
//
// FSM states: idle → awaiting_approval → approved / rejected → idle
package main

import (
	"encoding/json"
	"fmt"

	"github.com/bhatti/PlexSpaces/sdks/go/plexspaces"
)

// ApprovalGateActor manages human-in-the-loop approval for high-stakes agent actions.
//
// Agent calls: ask("approval_gate", "request_approval", {...})
// Then the agent suspends.
// A human (or automated policy) calls "approve" or "reject".
// The agent gets signaled and resumes.
//
// Key insight: the agent can wait for days. The DurabilityFacet preserves
// all state durably — no polling, no timeouts burning tokens.
type ApprovalGateActor struct {
	plexspaces.BaseActor
	ActorID         string         `json:"actor_id"`
	FSMState        string         `json:"fsm_state"`
	PendingRequest  map[string]any `json:"pending_request"`
	PendingAgentID  string         `json:"pending_agent_id"`
	DecisionHistory []map[string]any `json:"decision_history"`
}

func NewApprovalGateActor() plexspaces.Actor {
	a := &ApprovalGateActor{
		FSMState:        "idle",
		DecisionHistory: []map[string]any{},
		PendingRequest:  map[string]any{},
	}
	a.SetSelf(a)
	return a
}

func (ag *ApprovalGateActor) Init(configJSON string) string {
	var config struct {
		ActorID string `json:"actor_id"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	ag.SetRuntimeMetadata(config.ActorID)
	ag.ActorID = config.ActorID
	if ag.FSMState == "" {
		ag.FSMState = "idle"
	}
	if ag.DecisionHistory == nil {
		ag.DecisionHistory = []map[string]any{}
	}
	if ag.PendingRequest == nil {
		ag.PendingRequest = map[string]any{}
	}
	if err := host.PG().Join("svc:approval_gate"); err != nil {
		host.Warn(fmt.Sprintf("ApprovalGateActor: failed to join svc:approval_gate: %v", err))
	}
	host.Info(fmt.Sprintf("ApprovalGateActor Init actor_id=%s state=%s", config.ActorID, ag.FSMState))
	return ""
}

func (ag *ApprovalGateActor) Handle(fromActor, msgType, payloadJSON string) string {
	p := parsePayload(payloadJSON)
	op := stringVal(p, "op", msgType)
	switch op {
	case "request_approval":
		return ag.requestApproval(p)
	case "approve":
		return ag.approve(p)
	case "reject":
		return ag.reject(p)
	case "get_status":
		return ag.getStatus()
	case "get_history":
		return ag.getHistory()
	default:
		return marshal(map[string]any{"error": "unknown_op", "op": op})
	}
}

func (ag *ApprovalGateActor) requestApproval(p map[string]any) string {
	if ag.FSMState != "idle" {
		return marshal(map[string]any{
			"status":        "busy",
			"message":       fmt.Sprintf("Approval gate is already %s", ag.FSMState),
			"current_agent": ag.PendingAgentID,
		})
	}

	agentID := stringVal(p, "agent_id", "")
	action := stringVal(p, "action", "")
	context, _ := p["context"].(map[string]any)
	if context == nil {
		context = map[string]any{}
	}

	ag.FSMState = "awaiting_approval"
	ag.PendingAgentID = agentID
	ag.PendingRequest = map[string]any{
		"action":            action,
		"context":           context,
		"requested_at_ms":   host.NowMs(),
	}

	host.Info(fmt.Sprintf("ApprovalGate: request from agent=%s action=%s", agentID, action))
	ag.IncrCounter(host, "approval_requests_total")

	// Store request for external review
	reqJSON, _ := json.Marshal(map[string]any{
		"action":          action,
		"context":         context,
		"requested_at_ms": host.NowMs(),
		"agent_id":        agentID,
	})
	host.KVPut("approval_request:"+ag.ActorID, string(reqJSON))

	return marshal(map[string]any{
		"status":  "pending",
		"message": "Approval request submitted. Agent will be notified on decision.",
		"gate_id": ag.ActorID,
	})
}

func (ag *ApprovalGateActor) approve(p map[string]any) string {
	if ag.FSMState != "awaiting_approval" {
		return marshal(map[string]any{
			"error": fmt.Sprintf("No pending approval request (state=%s)", ag.FSMState),
		})
	}

	agentID := ag.PendingAgentID
	approver := stringVal(p, "approver", "")
	comment := stringVal(p, "comment", "")

	ag.FSMState = "approved"

	if ag.DecisionHistory == nil {
		ag.DecisionHistory = []map[string]any{}
	}
	ag.DecisionHistory = append(ag.DecisionHistory, map[string]any{
		"action":        stringVal(ag.PendingRequest, "action", ""),
		"decision":      "approved",
		"approver":      approver,
		"comment":       comment,
		"decided_at_ms": host.NowMs(),
	})

	// Signal the suspended agent to resume
	if agentID != "" {
		if result := host.Send(agentID, "workflow_signal:resume", map[string]any{
			"decision": "approved",
			"approver": approver,
			"comment":  comment,
		}); result != "" {
			host.Warn(fmt.Sprintf("ApprovalGate: failed to signal agent %s: %s", agentID, result))
		}
	}

	ag.FSMState = "idle"
	ag.PendingAgentID = ""
	ag.PendingRequest = map[string]any{}

	ag.IncrCounter(host, "approvals_granted_total")
	host.Info(fmt.Sprintf("ApprovalGate: approved agent=%s approver=%s", agentID, approver))

	return marshal(map[string]any{
		"status":   "approved",
		"agent_id": agentID,
		"approver": approver,
	})
}

func (ag *ApprovalGateActor) reject(p map[string]any) string {
	if ag.FSMState != "awaiting_approval" {
		return marshal(map[string]any{
			"error": fmt.Sprintf("No pending approval request (state=%s)", ag.FSMState),
		})
	}

	agentID := ag.PendingAgentID
	approver := stringVal(p, "approver", "")
	reason := stringVal(p, "reason", "")

	ag.FSMState = "rejected"

	if ag.DecisionHistory == nil {
		ag.DecisionHistory = []map[string]any{}
	}
	ag.DecisionHistory = append(ag.DecisionHistory, map[string]any{
		"action":        stringVal(ag.PendingRequest, "action", ""),
		"decision":      "rejected",
		"approver":      approver,
		"reason":        reason,
		"decided_at_ms": host.NowMs(),
	})

	// Signal agent with rejection
	if agentID != "" {
		if result := host.Send(agentID, "workflow_signal:resume", map[string]any{
			"decision": "rejected",
			"approver": approver,
			"reason":   reason,
		}); result != "" {
			host.Warn(fmt.Sprintf("ApprovalGate: failed to signal agent %s with rejection: %s", agentID, result))
		}
	}

	ag.FSMState = "idle"
	ag.PendingAgentID = ""
	ag.PendingRequest = map[string]any{}

	ag.IncrCounter(host, "approvals_rejected_total")
	host.Info(fmt.Sprintf("ApprovalGate: rejected agent=%s reason=%s", agentID, reason))

	return marshal(map[string]any{
		"status":   "rejected",
		"agent_id": agentID,
		"reason":   reason,
	})
}

func (ag *ApprovalGateActor) getStatus() string {
	return marshal(map[string]any{
		"status":           "ok",
		"state":            ag.FSMState,
		"pending_agent_id": ag.PendingAgentID,
		"pending_request":  ag.PendingRequest,
		"decision_count":   len(ag.DecisionHistory),
	})
}

func (ag *ApprovalGateActor) getHistory() string {
	if ag.DecisionHistory == nil {
		ag.DecisionHistory = []map[string]any{}
	}
	return marshal(map[string]any{
		"status":    "ok",
		"decisions": ag.DecisionHistory,
		"count":     len(ag.DecisionHistory),
	})
}

File: examples/python/apps/minipi/approval_gate.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""ApprovalGateActor — human-in-the-loop state machine.

Demonstrates: GenFSM for human-in-the-loop, durable wait (agent can be
suspended for days), and PlexSpaces's unique capability to resume exactly
where it left off without losing state.

FSM states: idle → awaiting_approval → approved / rejected → idle
"""

import json
from plexspaces import fsm_actor, state, init_handler, handler, host


@fsm_actor(states=["idle", "awaiting_approval", "approved", "rejected"], initial="idle")
class ApprovalGateActor:
    """
    Human-in-the-loop approval gate.

    Agent calls: host.ask("approval_gate", "request_approval", {...})
    Then the agent suspends (self.suspend()).
    A human (or automated policy) calls "approve" or "reject".
    The agent gets signaled and resumes.

    Key insight: the agent can wait for days. The DurabilityFacet preserves
    all state durably — no polling, no timeouts burning tokens.
    """

    actor_id: str = state(default="")
    fsm_state: str = state(default="idle")
    pending_request: dict = state(default_factory=dict)
    pending_agent_id: str = state(default="")
    decision_history: list = state(default_factory=list)

    @init_handler
    def on_init(self, config: dict) -> None:
        self.actor_id = config.get("actor_id", "")
        try:
            host.kv.put("svc:approval_gate", host.self_id())
        except Exception:
            pass
        try:
            host.registry.register(None, self.actor_id or host.self_id(), "actor", "",
                                   object_category="approval_gate")
        except Exception:
            pass
        host.info(f"ApprovalGateActor init actor_id={self.actor_id}")

    @handler("request_approval")
    def request_approval(self, agent_id: str = "", action: str = "", context: dict = None) -> dict:
        """An agent requests human approval for a high-stakes action."""
        if self.fsm_state != "idle":
            return {
                "status": "busy",
                "message": f"Approval gate is already {self.fsm_state}",
                "current_agent": self.pending_agent_id,
            }

        self.fsm_state = "awaiting_approval"
        self.pending_agent_id = agent_id
        self.pending_request = {
            "action": action,
            "context": context or {},
            "requested_at_ms": host.now_ms(),
        }

        host.info(f"ApprovalGate: request from agent={agent_id} action={action}")
        host.incr_counter("approval_requests_total", 1)

        # Store request for external review (e.g., dashboard display)
        try:
            host.kv.put(
                f"approval_request:{self.actor_id}",
                json.dumps({**self.pending_request, "agent_id": agent_id})
            )
        except Exception:
            pass

        return {
            "status": "pending",
            "message": "Approval request submitted. Agent will be notified on decision.",
            "gate_id": self.actor_id,
        }

    @handler("approve")
    def approve(self, approver: str = "", comment: str = "") -> dict:
        """Human approves the pending action."""
        if self.fsm_state != "awaiting_approval":
            return {"error": f"No pending approval request (state={self.fsm_state})"}

        agent_id = self.pending_agent_id
        self.fsm_state = "approved"

        self.decision_history.append({
            "action": self.pending_request.get("action", ""),
            "decision": "approved",
            "approver": approver,
            "comment": comment,
            "decided_at_ms": host.now_ms(),
        })

        # Signal the suspended agent to resume
        try:
            host.send(agent_id, "workflow_signal:resume", {
                "decision": "approved",
                "approver": approver,
                "comment": comment,
            })
        except Exception as e:
            host.warn(f"Failed to signal agent {agent_id}: {e}")

        self.fsm_state = "idle"
        self.pending_agent_id = ""
        self.pending_request = {}

        host.incr_counter("approvals_granted_total", 1)
        host.info(f"ApprovalGate: approved agent={agent_id} approver={approver}")

        return {"status": "approved", "agent_id": agent_id, "approver": approver}

    @handler("reject")
    def reject(self, approver: str = "", reason: str = "") -> dict:
        """Human rejects the pending action."""
        if self.fsm_state != "awaiting_approval":
            return {"error": f"No pending approval request (state={self.fsm_state})"}

        agent_id = self.pending_agent_id
        self.fsm_state = "rejected"

        self.decision_history.append({
            "action": self.pending_request.get("action", ""),
            "decision": "rejected",
            "approver": approver,
            "reason": reason,
            "decided_at_ms": host.now_ms(),
        })

        # Signal agent with rejection
        try:
            host.send(agent_id, "workflow_signal:resume", {
                "decision": "rejected",
                "approver": approver,
                "reason": reason,
            })
        except Exception as e:
            host.warn(f"Failed to signal agent {agent_id} with rejection: {e}")

        self.fsm_state = "idle"
        self.pending_agent_id = ""
        self.pending_request = {}

        host.incr_counter("approvals_rejected_total", 1)
        host.info(f"ApprovalGate: rejected agent={agent_id} reason={reason}")

        return {"status": "rejected", "agent_id": agent_id, "reason": reason}

    @handler("get_status")
    def get_status(self) -> dict:
        return {
            "status": "ok",
            "state": self.fsm_state,
            "pending_agent_id": self.pending_agent_id,
            "pending_request": self.pending_request,
            "decision_count": len(self.decision_history),
        }

    @handler("get_history")
    def get_history(self) -> dict:
        return {
            "status": "ok",
            "decisions": self.decision_history,
            "count": len(self.decision_history),
        }
